bootstrap.rkt

```diff
@@ -0,0 +1,476 @@
+#lang racket
+
+(require (rename-in racket [syntax-case racket-syntax-case]))
+(require racket/pretty)
+
+; Problem: bootstrapping can be difficult. When we don't yet have convenient
+; macro-defining macros like fancy-syntax-case, it can be inconvenient to write
+; complex macros such as fancy-syntax-case.
+;
+; Solution: instead of manually defining Klister's fancy-syntax-case using
+; Klister's more primitive raw-syntax-case, we write some Racket code which
+; expands to the code we would have written manually. This is easier, because
+; Racket does have convenient syntax-manipulating macros like match,
+; quasiquote, and racket-syntax-case.
+
+; (generate-syntax-case 'my-macro 'stx (list 'keyword)
+;   (list
+;     (list '()
+;           'rhs1)
+;     (list '((,a ,b) (,c ,d))
+;           'rhs2)
+;     (list '(keyword ,tail ...)
+;           '(> (length tail) 2)
+;           'rhs3)))
+; =>
+; '(let [failure-cc
+;        (lambda ()
+;          (syntax-error '"my-macro call has invalid syntax" stx))]
+;    (let [failure-cc
+;          (lambda ()
+;            (raw-syntax-case stx
+;              [(cons head tail)
+;               (if (identifier? head)
+;                   (>>= (free-identifier=? head 'keyword)
+;                     (lambda (same-identifier)
+;                       (if same-identifier
+;                         (if (> (length tail) 2)
+;                             rhs3
+;                             (failure-cc))
+;                         (failure-cc))))
+;                   (failure-cc))]
+;              [_ (failure-cc)]))]
+;      (let [failure-cc
+;            (lambda ()
+;              (raw-syntax-case stx
+;                [(cons ab cd-nil)
+;                 (raw-syntax-case ab
+;                   [(cons a b-nil)
+;                    (...etc... rhs2)]
+;                   [_ (failure-cc)])]
+;                [_ (failure-cc)]))]
+;        (let [failure-cc
+;              (lambda ()
+;                (raw-syntax-case stx
+;                  [() rhs1]
+;                  [_ (failure-cc)]))]
+;          (failure-cc)))))
+(define (generate-syntax-case macro-name stx-expr keywords cases)
+  (letrec ([stx-name (gensym 'stx-)]
```

Comment:
I have to use `gensym` a lot. One reason is that when the generated code is written to a file, only the names of the symbols are written; if there was any in-memory information about scope, such as which macro invocation introduced which name, that information is lost.